packages/cli/src/commands/governance/upvote.ts

```diff
@@ -29,6 +31,52 @@ export default class Upvote extends BaseCommand {
       .runChecks()
 
     const account = await (await this.kit.contracts.getAccounts()).voteSignerToAccount(signer)
-    await displaySendTx('upvoteTx', await governance.upvote(id, account), {}, 'ProposalUpvoted')
+
+    const consideredProposals = await this.dequeueAllPossibleProposals(governance as any)
+
+    if (!consideredProposals.some((k) => k.id === id)) {
+      await displaySendTx('upvoteTx', await governance.upvote(id, account), {}, 'ProposalUpvoted')
+    } else {
+      console.info(chalk.green('Proposal was dequeued, no need to upvote it.'))
+    }
+  }
+
+  /**
+   * Dequeues all possible proposals, returns the ones that were considered to be dequeued.
+   * This cycle is there only for safety purposes (in 99.9% of times it will not be run more than once).
+   * The reason behind this cycle is as follows:
+   * 1. Someone will DDoS our proposals (eg 100+)
+   * 2. All these proposals will expire
+   * 3. DequeueProposalsIfReady will always try to dequeue first n proposals but if all of them will be expired it will just remove them from queue.
+   * 4. Since none of the proposals were actually dequeued, next call will allow to dequeue again
+   * 5. Upvote function will try to dequeue again and possibly it will hit the proposal and bug that we have
+   */
+  async dequeueAllPossibleProposals(governance: GovernanceWrapper) {
+    const concurrentProposals = (await governance.concurrentProposals()).toNumber()
+    const queue = await governance.getQueue()
+    const originalLastDequeue = await governance.lastDequeue()
+
+    let consideredProposals
+
+    for (let index = 0; index < queue.length / concurrentProposals + 1; index++) {
+      consideredProposals = (
+        await Promise.all(
+          queue
+            .slice(0, concurrentProposals)
```

Comment:
trying to understand this. 

first run `concurrentProposals` would be `undefined`. second run it would be array of non expired proposals. 

but passing an array to slice like `[1,2,3,4].slice(0, [1,2,3])` results in `[]`

i think this line should be 

```suggestion
            .slice(0, concurrentProposals?.length)
```

Reply:
has this been tested for the edge case it was written for?

Reply:
Hey, 
`concurrentProposals` variable is of type number (it is not array). I introduced tests in upvote.test.ts.

Reply:
oh bad reading i mixed up considered and current

nit: call it `currentProposalCount`